internal/settings/config_provider.go

```diff
@@ -0,0 +1,399 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package settings
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/uber/prototool/internal/x/strs"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v2"
+)
+
+type configProvider struct {
+	logger                   *zap.Logger
+	filePathToConfig         map[string]Config
+	dirPathToFilePath        map[string]string
+	dirPathToExcludePrefixes map[string][]string
+	lock                     sync.RWMutex
```

Comment:
There is a lot of logic written around making the `configProvider` concurrent-safe. Is this really necessary? In the places that this might be called concurrently, such as in calls in `internal/x/file`, can we instead synchronously run the commands? 

I'm not sure how much of a performance gain we expect to see in this domain, so it might be worth removing it for now. 

What do you think? Is this actually completely necessary / am I missing something?

Reply:
Things like caches (which is what this is locking) should always be thread-safe, IMO, I don't see it as an optional thing. I don't want our designs to have to rely on something being not thread-safe, or worse, someone using this in a deep dependency in a threaded manner without realizing this is not thread-safe. It's defensive programming if anything, but a cache should generally always be thread-safe unless there's a compelling reason not to be IMO.